util/mon: pass reserved account by reference

yuzefovich opened the pull request:
**util/mon: pass reserved account by reference**

This commit makes it so that we now pass the `reserved` memory account
when starting up memory monitors by reference. Previously, due to
passing by value, when the monitor is stopped, the copy of the value
would get used so that the actual "reserved" memory account would get
out of sync with its user. This is now fixed. However, this bug doesn't
really have any production impact since we use this "reserved" feature
in a handful of places and these "reserved" memory accounts are not
reused between different usages.

Additionally, this commit renames `MakeStandaloneBudget` to
`NewStandaloneBudget` (since it now returns a reference) and adds
a separate "start" method when the caller doesn't want to pre-reserve
anything when starting up a monitor.

Release note: None

**sql: make sure to close the reserved account for each session**

This commit makes it more clear that the "reserved" memory account
created for each connection is closed when that connection is closed.
Previously, the account was already cleared (when "session root" monitor
is stopped which happens when the connExecutor stops), so there was no
leak in the accounting system because of it, yet it wasn't obvious - this
commit makes it more obvious.

Release note: None

yuzefovich committed:
sql: make sure to close the reserved account for each session

This commit makes it more clear that the "reserved" memory account
created for each connection is closed when that connection is closed.
Previously, the account was already cleared (when "session root" monitor
is stopped which happens when the connExecutor stops), so there was no
leak in the accounting system because of it, yet it wasn't obvious - this
commit makes it more obvious.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/83629)
<!-- Reviewable:end -->

yuzefovich committed:
util/mon: pass reserved account by reference

This commit makes it so that we now pass the `reserved` memory account
when starting up memory monitors by reference. Previously, due to
passing by value, when the monitor is stopped, the copy of the value
would get used so that the actual "reserved" memory account would get
out of sync with its user. This is now fixed. However, this bug doesn't
really have any production impact since we use this "reserved" feature
in a handful of places and these "reserved" memory accounts are not
reused between different usages.

Additionally, this commit renames `MakeStandaloneBudget` to
`NewStandaloneBudget` (since it now returns a reference) and adds
a separate "start" method when the caller doesn't want to pre-reserve
anything when starting up a monitor.

Release note: None

knz reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> with tip

> Reviewed 57 of 57 files at r1, 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83629#-:-N6FefEoBLD6pbgwNV7p:b-xv605s)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @michae2 and @yuzefovich)

___
*[`pkg/sql/conn_executor.go` line 760 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/83629#-N6FdylFCETDVqitZHBZ:-N6FdylFCETDVqitZHB_:b-wih7qw) ([raw file](https://github.com/cockroachdb/cockroach/blob/581c07d7e0ccf83d068b3ce6d0c187c6ff6fe01f/pkg/sql/conn_executor.go#L760)):*
> ```Go
> 		// also cannot close it before closeWrapper since we need to close the
> 		// internal monitors of the connExecutor first.
> 		defer reserved.Close(ctx)
> ```

1) if you pull the `defer` out of the closure, you'll save some bytes.
```go
defer reserved.Close(ctx)
defer func() { ... }()
return h.ex.run
```

2) you can save the entire allocation by passing the outer variables as arg:
```go
defer func(ctx context.Context, h ConnectionHandler) {...} (ctx, h)
```


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83629#-:-N6Fr9949fY5btniiTK5:b-oduykh)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @knz and @michae2)

___
*[`pkg/sql/conn_executor.go` line 760 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/83629#-N6FdylFCETDVqitZHBZ:-N6Fr1qH2hTF-KMmRBWb:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/581c07d7e0ccf83d068b3ce6d0c187c6ff6fe01f/pkg/sql/conn_executor.go#L760)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

1) if you pull the `defer` out of the closure, you'll save some bytes.
```go
defer reserved.Close(ctx)
defer func() { ... }()
return h.ex.run
```

2) you can save the entire allocation by passing the outer variables as arg:
```go
defer func(ctx context.Context, h ConnectionHandler) {...} (ctx, h)
```
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->
